### PR TITLE
fix scheduler queue update logic

### LINKED ISSUE #98828: Scheduler PriorityQueue Update function
**What happened**:

When we find the pod in podBackoffQ, we delete it from podBackoffQ and add it to the active queue directly, which doesn't match the comment.
Also, when we find the pod in unschedulableQ, we don't respect the backoff policy, we delete the pod from unschedulableQ and add it to the active queue directly.

**What you expected to happen**:

If the pod is in the backoff queue, update it there.
Also, when the pod is in the unschedulable queue, respect the backoff policy too.

**Anything else we need to know?**:

Not sure if this works as expected or is a bug, i guess it is a bug, so i send a PR for this.

### PR DESCRIPTION
**Pull request by NickrenREN**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Part of #98828

respect pod backoff policy when UpdatePod

**Special notes for your reviewer**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 1]
```diff
@@ -444,12 +440,24 @@ func (p *PriorityQueue) Update(oldPod, newPod *v1.Pod) error {
...
+				if err := p.podBackoffQ.Add(updatePod(usPodInfo, newPod)); err != nil {
+					klog.Errorf("Error adding pod %v to the backoff queue: %v", usPodInfo.Pod.Name, err)
+					return err
+				} else {
+					p.unschedulableQ.delete(usPodInfo.Pod)
+					return nil
+				}
+			} else {
+				if err := p.activeQ.Add(updatePod(usPodInfo, newPod)); err != nil {
+					klog.Errorf("Error adding pod %v to the scheduling queue: %v", usPodInfo.Pod.Name, err)
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue.go:453]: this error is already logged in the caller of this function

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 2]
```diff
@@ -426,29 +426,30 @@ func (p *PriorityQueue) Update(oldPod, newPod *v1.Pod) error {
...
-			pInfo := updatePod(usPodInfo, newPod)
-			p.PodNominator.UpdateNominatedPod(oldPod, pInfo.PodInfo)
-			if err := p.activeQ.Add(pInfo); err != nil {
-				return err
+			var uErr error
+			if p.isPodBackingoff(usPodInfo) {
+				if uErr = p.podBackoffQ.Add(pInfo); uErr == nil {
+					p.unschedulableQ.delete(usPodInfo.Pod)
+				}
+				return uErr
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue.go:443]:
```suggestion
				if err := p.podBackoffQ.Add(pInfo); err != nil {
					return err
				}
				p.unschedulableQ.delete(usPodInfo.Pod)
```

Do similarly below

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 3]
```diff
@@ -329,6 +330,10 @@ func TestPriorityQueue_Update(t *testing.T) {
 	}
 	updatedPod := medPriorityPodInfo.Pod.DeepCopy()
 	updatedPod.ClusterName = "test"
+
+	// Move clock by podInitialBackoffDuration, so that pods in the unschedulableQ would pass the backing off,
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:334]: You should unit test both cases. So do one scenario where the pod goes to backoff queue, and one where it goes directly to active.

We also need a test for the pod is in backoff queue and stays there after an update.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 4]
```diff
@@ -321,14 +322,46 @@ func TestPriorityQueue_Update(t *testing.T) {
...
 	updatedPod.ClusterName = "test"
+
+	q.Update(medPriorityPodInfo.Pod, updatedPod)
+	rawPodInfo, err = q.podBackoffQ.Pop()
+	podGotFromBackoffQ = rawPodInfo.(*framework.QueuedPodInfo).Pod
+	if err != nil || podGotFromBackoffQ != updatedPod {
+		t.Errorf("Expected: %v after Pop, but got: %v", updatedPod.Name, podGotFromBackoffQ.Name)
+	}
+
+	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(medPriorityPodInfo.Pod), q.SchedulingCycle())
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:356]: Add a comment above describing the case, as you did for the others.

Even better, can you call `t.Run` for each testcase? They are independent from each other after all.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 5]
```diff
@@ -321,14 +322,46 @@ func TestPriorityQueue_Update(t *testing.T) {
...
+
+	// updating a pod which is in unschedulable queue, and it is still backing off,
+	// we will move it to backoff queue
 	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(medPriorityPodInfo.Pod), q.SchedulingCycle())
 	if len(q.unschedulableQ.podInfoMap) != 1 {
 		t.Error("Expected unschedulableQ to be 1.")
 	}
 	updatedPod := medPriorityPodInfo.Pod.DeepCopy()
 	updatedPod.ClusterName = "test"
+
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:348]: remove this empty line so we know it semantically belongs to the above section.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 6]
```diff
@@ -321,14 +322,46 @@ func TestPriorityQueue_Update(t *testing.T) {
 	if p, err := q.Pop(); err != nil || p.Pod != highPriNominatedPodInfo.Pod {
 		t.Errorf("Expected: %v after Pop, but got: %v", highPriorityPodInfo.Pod.Name, p.Pod.Name)
 	}
-	// Updating a pod that is in unschedulableQ in a way that it may
-	// become schedulable should add the pod to the activeQ.
+
+	// Updating a pod that is in backoff queue and it is still backing off
+	// pod will not be moved to active queue, and it will be updated in backoff queue
+	podInfo := q.newQueuedPodInfo(medPriorityPodInfo.Pod)
+	podInfo.Attempts = 2
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:329]: nit: it's not needed, right?

**Review by Huang-Wei — COMMENTED** [review]
Some nits.

**Comment by Huang-Wei** [comment]
@NickrenREN could you rebase and take a look latest comments?

**Comment by NickrenREN** [comment]
@Huang-Wei done

**Comment by alculquicondor** [comment]
Can you add a description of what the update is about? We should be diligent with PR descriptions so that it's easy to navigate through code changes

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 4 above)
Comment (alculquicondor) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:358]: Please do a follow up where each "test case" is wrapped in a `t.Run`

**Review by alculquicondor — COMMENTED** [review]
Please squash

/approve
/priority important-soon

**Comment by ahg-g** [comment]
Please update the description to say "Part of #98828" instead of fixes since we still want to ignore updates to assumed pods.

/hold

**Comment by NickrenREN** [comment]
squashed and updated......
PTAL

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 7]
```diff
@@ -323,15 +324,47 @@ func TestPriorityQueue_Update(t *testing.T) {
...
+	// updating a pod which is in unschedulable queue, and it is not backing off,
+	// we will move it to active queue
 	q.AddUnschedulableIfNotPresent(q.newQueuedPodInfo(medPriorityPodInfo.Pod), q.SchedulingCycle())
 	if len(q.unschedulableQ.podInfoMap) != 1 {
 		t.Error("Expected unschedulableQ to be 1.")
 	}
 	updatedPod := medPriorityPodInfo.Pod.DeepCopy()
 	updatedPod.ClusterName = "test"
 	q.Update(medPriorityPodInfo.Pod, updatedPod)
+	rawPodInfo, err = q.podBackoffQ.Pop()
```
Comment (Huang-Wei) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:350]: Shoulnd't we pop from activeQ in this cycle?

Comment (NickrenREN) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:350]: sorry for the misleading comments, updated

Comment (Huang-Wei) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:350]: LGTM now.

**Comment by alculquicondor** [comment]
Can this be considered a critical bug?

**Pull request merged**